### PR TITLE
XIVY-12797 Impl. of expandable Code Editor

### PR DESCRIPTION
**Pull request by ivy-edp**
I have now created the expandable code editor. I have already integrated it into the StartPart and TaskPart. However, before incorporating it into all other locations (approximately 7 additional places where the ScriptArea is used), I would like to hear your opinion on whether the integration in these two parts is okay. Even though we discussed earlier that I should leave the ScriptArea component as it is and not add additional parameters and if/else statements, it might now make sense to incorporate it directly there. This way, I could avoid code repetition in too many places. Additionally, I believe that the ScriptArea is only used in CodeParts, so the MaximizedCodeEditorBrowser always needs to be integrated with a ScriptArea. However, I might be mistaken, and it could be better to incorporate everything as shown in the StartPart and TaskPart, keeping the ScriptArea doing only what it was originally intended for. What do you think?

grafik
grafik

### CODE REVIEW

File: packages/editor/src/components/browser/MaximizedCodeEditorBrowser.tsx [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+  useEffect(() => {
+    const handleResize = () => {
+      setCodeEditorHeight(window.innerHeight - 150);
+    };
+    window.addEventListener('resize', handleResize);
+    return () => {
+      window.removeEventListener('resize', handleResize);
+    };
+  }, []);
```
Comment (ivy-lli) [inline packages/editor/src/components/browser/MaximizedCodeEditorBrowser.tsx:43]: pretty sure this is not needed

File: packages/editor/src/components/browser/MaximizedCodeEditorBrowser.tsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+              value={editorValue}
+              onChange={applyEditor}
+              context={{ location }}
+              height={codeEditorHeight}
+              onMountFuncs={[setEditor]}
+              options={MAXIMIZED_MONACO_OPTIONS}
+            />
+          </div>
+          <Browser {...browser} types={browsers} accept={modifyEditor} location={location} initSearchFilter={getMonacoSelection} />
+        </div>
```
Comment (ivy-lli) [inline packages/editor/src/components/browser/MaximizedCodeEditorBrowser.tsx:60]: make own component

File: packages/editor/src/components/browser/useMaximizedCodeEditor.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+  const maximizeCode: FieldsetControl = {
+    label: 'Fullsize Code Editor',
+    icon: IvyIcons.ArrowsMaximize,
+    action: () => {
+      setIsMaximizedCodeEditorOpen(true);
+    }
+  };
+
+  return { isMaximizedCodeEditorOpen, maximizeCode, setIsMaximizedCodeEditorOpen };
```
Comment (ivy-lli) [inline packages/editor/src/components/browser/useMaximizedCodeEditor.ts:16]:
```
return { openState: {isMaximizedCodeEditorOpen, setIsMaximizedCodeEditorOpen}, maximizeCode };
```

**Review by ivy-lli — CHANGES_REQUESTED** [review]

File: packages/editor/src/components/browser/MaximizedCodeEditorBrowser.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      icon: IvyIcons.StartProgram
+    }
+  ];
+
+  return (
+    <>
+      <Dialog open={open} onOpenChange={onOpenChange}>
+        <BrowserBody activeTab={'maxCode'} open={open} tabs={tabs} />
+      </Dialog>
+    </>
```
Comment (ivy-lli) [inline packages/editor/src/components/browser/MaximizedCodeEditorBrowser.tsx:37]: fragment not needed

File: packages/editor/src/components/widgets/code-editor/ScriptArea.tsx [hunk 5]
```diff
@@ -4,13 +4,30 @@ import ResizableCodeEditor from './ResizableCodeEditor';
 import { Browser, useBrowser } from '../../../components/browser';
 import { useMonacoEditor } from './useCodeEditor';
 import { usePath } from '../../../context';
+import MaximizedCodeEditorBrowser from '../../browser/MaximizedCodeEditorBrowser';
 
-const ScriptArea = (props: CodeEditorAreaProps) => {
+type ScriptAreaProps = CodeEditorAreaProps & {
+  openState: {
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/code-editor/ScriptArea.tsx:10]: maybe maximizeState or something like that is a bit clearer for a user of the ScriptArea...

File: packages/editor/src/components/widgets/headlineControls/HeadlineControls.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import './HeadlineControls.css';
+import Button from '../button/Button';
+import type { FieldsetControl } from '../fieldset';
+
+type HeadlineControlsProps = {
+  controls: FieldsetControl[];
+  open?: boolean;
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/headlineControls/HeadlineControls.tsx:7]: the open has nothing to do with this component IMO. This is only a special case for the collapsible, that if it is open the controls are shown. So just handle it there.

File: packages/editor/src/monaco/monaco-editor-util.ts [hunk 7]
```diff
@@ -23,6 +23,17 @@ export const MONACO_OPTIONS: monaco.editor.IStandaloneEditorConstructionOptions
...
 };
 
+export const MAXIMIZED_MONACO_OPTIONS: monaco.editor.IStandaloneEditorConstructionOptions = {
+  ...MONACO_OPTIONS,
+  lineNumbers: 'on',
+  folding: true,
+  renderLineHighlight: 'gutter',
+  cursorStyle: 'line',
+  cursorBlinking: 'blink',
+  contextmenu: false, // Disabling context menu
```
Comment (ivy-lli) [inline packages/editor/src/monaco/monaco-editor-util.ts:33]: already in the default MONACO_OPTIONS

File: packages/editor/src/monaco/monaco-editor-util.ts [hunk 8]
```diff
@@ -23,6 +23,17 @@ export const MONACO_OPTIONS: monaco.editor.IStandaloneEditorConstructionOptions
   fixedOverflowWidgets: true
 };
 
+export const MAXIMIZED_MONACO_OPTIONS: monaco.editor.IStandaloneEditorConstructionOptions = {
+  ...MONACO_OPTIONS,
+  lineNumbers: 'on',
+  folding: true,
+  renderLineHighlight: 'gutter',
+  cursorStyle: 'line',
+  cursorBlinking: 'blink',
```
Comment (ivy-lli) [inline packages/editor/src/monaco/monaco-editor-util.ts:32]: isn't this the default?

File: packages/editor/src/components/widgets/collapsible/Collapsible.tsx [hunk 9]
```diff
@@ -22,14 +30,15 @@ const Collapsible = ({ label, defaultOpen, message, children, autoClosable }: Co
   }, [autoClosable, defaultOpen]);
   return (
     <CollapsibleRoot className='collapsible-root' open={open} onOpenChange={setOpen}>
-      <div className='collapsible-header'>
+      <div className={`collapsible-header${!open ? '-closed' : ''}`}>
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/collapsible/Collapsible.tsx:33]: I wouldn't change the whole class. Makes it only much harder to test IMO.
Better don't change the default component class, and handle such state stuff in a second class or in a data attribute e.g. `data-state={open ? 'open' : 'closed'}`

And if you do such things, I can only speak for myself, but IMO negative stuff is much harder to read. So even if you don't change this to the attribute style I would write it like `open ? '' : '-closed'`

**Review by ivy-lli — CHANGES_REQUESTED** [review]

**Comment by ivy-lli** [comment]
a lot better IMO. Some small stuff to improve :)

File: packages/editor/src/monaco/monaco-editor-util.ts (hunk 8 above)
Comment (ivy-edp) [reply packages/editor/src/monaco/monaco-editor-util.ts:32]: yes, you're right 👍

File: packages/editor/src/components/widgets/collapsible/Collapsible.css [hunk 10]
```diff
@@ -1,10 +1,10 @@
-.collapsible-root .collapsible-header {
+.collapsible-root .collapsible-header:has([data-state='open']) {
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/collapsible/Collapsible.css:1]: `:has` is not needed here
just do `.collapse-header[data-state='open']`

And I think you should have a default style without a data-state and then the style for a specific data state e.g. 
```css
.collapsible-root .collapsible-header {
  display: flex;
}
.collapsible-root .collapsible-header[data-state='open'] {
  justify-content: space-between;
  align-items: center;
}
```

Comment (ivy-lli) [reply packages/editor/src/components/widgets/collapsible/Collapsible.css:1]: Ah if you want to check the child state, yes then has is needed.
```css
.collapsible-root .collapsible-header {
  display: flex;
}
.collapsible-root .collapsible-header:has([data-state='open']) {
  justify-content: space-between;
  align-items: center;
}
```

**Review by ivy-lli — APPROVED** [review]

**Comment by ivy-lli** [comment]
Cool 👍

**Pull request merged**